linkerd/app/core/src/control.rs

```diff
@@ -145,160 +145,361 @@ pub mod add_origin {
-        dns: dns::Resolver,
+    #[derive(Clone)]
+    pub struct Resolve<S> {
+        dns: S,
     }
 
-    #[derive(Clone, Debug)]
-    pub struct Resolve<M> {
-        dns: dns::Resolver,
-        inner: M,
-    }
+    // === impl Resolve ===
 
-    #[pin_project]
-    pub struct Init<M>
-    where
-        M: tower::Service<client::Target>,
-    {
-        #[pin]
-        state: State<M>,
+    impl<S> Resolve<S> {
+        pub fn new(dns: S) -> Self {
+            Self { dns }
+        }
     }
 
-    #[pin_project(project = StateProj)]
-    enum State<M>
+    type UpdatesStream =
+        Pin<Box<dyn Stream<Item = Result<Update<Target>, Error>> + Send + 'static>>;
+
+    impl<S> tower::Service<ControlAddr> for Resolve<S>
     where
-        M: tower::Service<client::Target>,
+        S: tower::Service<dns::Name, Response = dns::ResolveResponse, Error = dns::Error>,
+        S: Clone + Send + 'static,
+        S::Future: Send,
     {
-        Resolve(#[pin] dns::IpAddrFuture, Option<(M, ControlAddr)>),
-        NotReady(M, Option<(SocketAddr, ControlAddr)>),
-        Inner(#[pin] M::Future),
-    }
+        type Response = resolve::FromStream<UpdatesStream>;
+        type Error = Error;
+        type Future = future::Ready<Result<Self::Response, Self::Error>>;
 
-    #[derive(Debug)]
-    pub enum Error<I> {
-        Dns(dns::Error),
-        Inner(I),
+        fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+            Poll::Ready(Ok(()))
+        }
+
+        fn call(&mut self, target: ControlAddr) -> Self::Future {
+            futures::future::ok(resolve::from_stream::<UpdatesStream>(Box::pin(
+                resolution_stream(self.dns.clone(), target.clone()),
+            )))
+        }
     }
 
-    // === impl Layer ===
+    // === impl ControlPlaneResolution ===
 
-    pub fn layer<M>(dns: dns::Resolver) -> impl svc::Layer<M, Service = Resolve<M>> + Clone
-    where
-        M: tower::Service<client::Target> + Clone,
-    {
-        svc::layer::mk(move |inner| Resolve {
-            dns: dns.clone(),
-            inner,
-        })
-    }
+    pub fn diff_targets(
+        old_targets: &Vec<Target>,
+        new_targets: &Vec<Target>,
+    ) -> (Vec<(SocketAddr, Target)>, Vec<SocketAddr>) {
+        let mut adds = Vec::default();
+        let mut removes = Vec::default();
 
-    // === impl Resolve ===
+        for new_target in new_targets {
+            if !old_targets.contains(new_target) {
+                adds.push((new_target.addr, new_target.clone()))
+            }
+        }
 
-    impl<M> tower::Service<ControlAddr> for Resolve<M>
+        for old_target in old_targets {
+            if !new_targets.contains(old_target) {
+                removes.push(old_target.addr)
+            }
+        }
+
+        (adds, removes)
+    }
+
+    pub fn resolution_stream<S>(
+        mut dns: S,
+        address: ControlAddr,
+    ) -> impl Stream<Item = Result<Update<Target>, Error>>
```

Comment:
[This comment](https://github.com/linkerd/linkerd2-proxy/pull/594#pullrequestreview-459533777) should have referenced this implementation